sdk/core/azure-core/src/http/policy_retry.cpp

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <http/policy.hpp>
+
+#include <algorithm>
+#include <cstdlib>
+#include <limits>
+#include <thread>
+
+using namespace Azure::Core::Http;
+
+namespace {
+decltype(RetryOptions::RetryDelay) ShouldRetry(
+    Response const& response,
+    RetryOptions const& retryOptions,
+    decltype(RetryOptions::MaxRetries) attempt)
+{
+  // Are we out of retry attempts?
+  if (attempt > retryOptions.MaxRetries)
+  {
+    return std::chrono::seconds(-1);
+  }
+
+  // Should we retry on the given response retry code?
+  auto const& statusCodes = retryOptions.StatusCodes;
+  auto const statusCodesEnd = statusCodes.end();
+  if (std::find(statusCodes.begin(), statusCodesEnd, response.GetStatusCode()) == statusCodesEnd)
+  {
+    return std::chrono::seconds(-1);
```

Comment:
So to my understanding, if the response code is not one of `retryOptions.StatusCodes`, we don't retry.

I don't know what other service sdk do. But storage SDK will retry if status code >= 500. Storage SDK will also retry on connection error and dns failure, in which cases there wouldn't be any status code because the connection isn't established at all.

Reply:
I'll check with Victor to see what would happen in case o DNS failure.
For status code 500, in the storage client, you can include 500 to options, and it will retry, no problem. I did not include 500, because I took the retry values from Go SDK.
I'll see what other SDKs do in case there's DNS failure. I can see arguments for both retry and an exception. Maybe we can add a value to options on what to do.

Reply:
I meant not only 500, for all status code >= 500,  we would retry. We cannot do that with current implementation.

Reply:
AFAIK, no other SDK would allow you to do it either, some of them don't even have StatusCodes configurable. Which doesn't mean that you don't need it and  we shouldn't think about adding it, but probably means that they've figured out how to do it without that - maybe they have custom implementations for some clients (such as storage hving their own retry policy implementation), I can see what they are doing. I don't think that `>=` logic realistically can be reasonably implemented implemented with  anything other than callback/overload/delegate/functor etc. But in my opinion there is still plenty of retry logic that is worth sharing - parsing retry headers, calculating exponential delay. If you need a thing like "retry on codes >= 500", we have to add a virtual function, something like `bool RetryOptions::ShouldRetry(HttpCode httpCode)` whic by default looks into StatusCodes set, but can be overridden to do the '>=' logic.

Reply:
virtual function sounds good.

Reply:
The Storage Go SDK retries on HTTP 500 and 503 status codes ONLY. A lot of discussion went into this and I'm confident that this is the right choice. We do not need to support any 5xx status code.

Reply:
Okay.

Reply:
How about network failure, dns failure and network interruption?

Reply:
Handling them now.